File: solution/0127____Word_Ladder.cpp
#include "util\frequently_included.h"

class Solution {
public:
        int ladderLength(string beginWord, string endWord, vector<string>& wordList) {
                return _1(beginWord, endWord, wordList);
        }
private:
        /**
         * @brief tle
         */
        int _1(string& begin, string& end, vector<string>& wordList){
                unordered_set<string> us;
                for(auto str: wordList)
                        us.insert(str);
                vector<string> prev = {begin};
                vector<string> next;
                int ans = 1;
                while(prev.size()){
                        ++ans;
                        for(auto& e: prev)
                                us.erase(e);
                        for(auto s: us)
                                for(auto ss: prev)
                                        if(_1_adjcent_words(s, ss)){
                                                next.push_back(s);
                                                if(s == end)
                                                        return ans;
                                        }
                        prev.swap(next);
                        next.clear();
                }
                return false;
        }
        bool _1_adjcent_words(string& s, string& ss){
                bool same = true;
                for(int i = 0; i < s.size(); ++i)
                        if(s[i] != ss[i])
                                if(same)
                                        same = false;
                                else
                                        return false;
                return true;
        }
};

int main(){
        vector<string> arr = {"hot","dot","dog","lot","log","cog"};
        Solution sln;
        sln.ladderLength("hit", "cog", arr);
        return 0;
}

File: solution/0127____Word Ladder.go
package main

func ladderLength(beginWord string, endWord string, wordList []string) int {
	return o127(beginWord, endWord, wordList)
}

// perf: 12, 95
func o127(beginWord string, endWord string, wordList []string) (ans int) {

	used := make([]bool, len(wordList), len(wordList))
	queue := []string{beginWord}

	for len(queue) > 0 {
		ans++
		var queue_temp []string
		for _, word := range queue {
			for i := range wordList {
				if used[i] {
					continue
				}
				if o127_helper(word, wordList[i]) {
					if wordList[i] == endWord {
						return ans + 1
					}
					queue_temp = append(queue_temp, wordList[i])
					used[i] = true
				}
			}
		}
		queue = queue_temp
	}
	return 0
}

func o127_helper(s, ss string) bool {
	i := 0
	for ; i < len(s); i++ {
		if s[i] != ss[i] {
			break
		}
	}
	if i == len(s) {
		// s == ss
		return false
	}
	i++
	for ; i < len(s); i++ {
		if s[i] != ss[i] {
			break
		}
	}
	return i == len(s)
}

func main() {

}
